pkg/kapp/diff/explicit_versioned_ref.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package diff
+
+import (
+	"fmt"
+)
+
+type VersionedRefDesc struct {
+	Namespace string `json:"namespace"`
+	APIGroup  string `json:"apiGroup"`
+	Kind      string `json:"kind"`
+	Name      string `json:"name"`
+}
+
+type ExplicitVersionedRefAnn struct {
+	References     []VersionedRefDesc `json:"references"`
+	VersionedNames map[string]string
+}
+
+type ExplicitVersionedRef struct {
+	Resource   VersionedResource
+	Annotation ExplicitVersionedRefAnn
+}
+
+const (
+	explicitReferenceKey = "kapp.k14s.io/versioned-explicit-ref"
+)
+
+func NewExplicitVersionedRef(res VersionedResource, annotation ExplicitVersionedRefAnn) *ExplicitVersionedRef {
+	return &ExplicitVersionedRef{res, annotation}
+}
+
+// Returns true if the
+func (e *ExplicitVersionedRef) IsReferenced() (bool, error) {
+	references, err := e.refStringList()
+	if err != nil {
+		return false, err
+	}
+
+	referenceKey := e.resourceRefString()
+
+	for _, v := range references {
+		if v == referenceKey {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (e *ExplicitVersionedRef) refStringList() ([]string, error) {
+	list := []string{}
+	for _, v := range e.Annotation.References {
+		v, err := e.validateAndReplaceCoreGroup(v)
+		if err != nil {
+			return list, err
+		}
+
+		list = append(list, fmt.Sprintf("%s/%s/%s/%s", v.Namespace, v.APIGroup, v.Kind, v.Name))
+	}
+	return list, nil
+}
+
+func (e *ExplicitVersionedRef) validateAndReplaceCoreGroup(resourceDescription VersionedRefDesc) (VersionedRefDesc, error) {
```

Comment:
Is this too much overhead for an advanced feature?

Reply:
I feel that the validation part is bit too much, as anyone using kapp would know that these are the basic things required to identify a resource. Although, I think that having it is not a bad idea either.

Reply:
Agreed. I guess I will leave as is for now